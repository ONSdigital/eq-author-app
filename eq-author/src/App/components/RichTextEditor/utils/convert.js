import { convertToHTML, convertFromHTML } from "draft-convert";
import { EditorState } from "draft-js";

export const toHTML = entityMap => {
  const entityToHTML = (entity, originalText) => {
    const mapper = entityMap[entity.type];
    return mapper ? mapper(entity) : originalText;
  };

  const convert = convertToHTML({ entityToHTML });

  return editorState => convert(editorState.getCurrentContent());
};

export const fromHTML = htmlToEntity => {
  const convert = convertFromHTML({ htmlToEntity });

  return (html, decorator) =>
    EditorState.createWithContent(convert(html), decorator);
};
