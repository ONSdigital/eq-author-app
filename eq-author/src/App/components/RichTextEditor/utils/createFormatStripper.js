import { mapKeys } from "lodash";
import { filterEditorState } from "draftjs-filters";
import { filterConfig } from "App/components/RichTextEditor/entities/PipedValue";

const mapper = {
  bold: { format: "BOLD", type: "styles" },
  emphasis: { format: "ITALIC", type: "styles" },
  list: { format: "unordered-list-item", type: "blocks" },
  heading: { format: "header-two", type: "blocks" }
};

export default function createFormatStripper(controls) {
  const filterConfiguration = {
    blocks: [],
    styles: [],
    entities: [filterConfig],
    maxNesting: 0,
    whitespacedCharacters: []
  };

  mapKeys(controls, (value, key) => {
    if (mapper[key] && controls[key]) {
      filterConfiguration[mapper[key].type].push(mapper[key].format);
    }
  });

  return editorState => filterEditorState(filterConfiguration, editorState);
}
