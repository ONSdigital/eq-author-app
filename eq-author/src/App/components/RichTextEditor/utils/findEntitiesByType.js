export default entityType => {
  return function findEntitiesByType(contentBlock, callback, contentState) {
    contentBlock.findEntityRanges(character => {
      const entityKey = character.getEntity();
      return (
        character.getEntity() !== null &&
        contentState.getEntity(entityKey).getType() === entityType
      );
    }, callback);
  };
};
