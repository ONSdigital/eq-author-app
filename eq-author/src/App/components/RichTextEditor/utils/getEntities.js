function getEntities(contentState, entityType) {
  const entities = [];

  contentState.getBlocksAsArray().forEach(block => {
    let selectedEntity = null;

    block.findEntityRanges(
      character => {
        const entityKey = character.getEntity();

        if (entityKey === null) {
          return false;
        }

        const entity = contentState.getEntity(entityKey);

        if (entityType && entity.getType() !== entityType) {
          return false;
        }

        selectedEntity = {
          entityKey,
          entity,
          blockKey: block.getKey()
        };

        return true;
      },
      (start, end) => {
        entities.push({ ...selectedEntity, start, end });
      }
    );
  });

  return entities;
}

export default getEntities;
