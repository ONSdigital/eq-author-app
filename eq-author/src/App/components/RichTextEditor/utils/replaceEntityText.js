import { SelectionState, Modifier } from "draft-js";

const getEntityRange = (block, entityKey) => {
  let range;

  // get the range for the entity key
  block.findEntityRanges(
    char => char.getEntity() === entityKey,
    (start, end) => {
      range = { start, end };
    }
  );

  return SelectionState.createEmpty(block.key).merge({
    anchorKey: block.key,
    anchorOffset: range.start,
    focusKey: block.key,
    focusOffset: range.end
  });
};

const replaceText = (contentState, entityKey, blockKey, text, inlineStyles) => {
  const block = contentState.getBlockForKey(blockKey);
  const entitySelection = getEntityRange(block, entityKey);

  // first replace text
  const updatedContentState = Modifier.replaceText(
    contentState,
    entitySelection,
    text,
    inlineStyles,
    entityKey
  );

  // then create new selection, which covers the updated entity
  const updatedEntitySelection = entitySelection.merge({
    focusOffset: entitySelection.getStartOffset() + text.length
  });

  // apply the updated entity to the content
  return Modifier.applyEntity(
    updatedContentState,
    updatedEntitySelection,
    entityKey
  );
};

export default replaceText;
