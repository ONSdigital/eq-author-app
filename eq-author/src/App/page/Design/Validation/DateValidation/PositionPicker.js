import React from "react";
import PropTypes from "prop-types";
import { RelativePositionSelect } from "./components";

const RELATIVE_POSITIONS = ["before", "after"];

const PositionPicker = ({value, onChange, onUpdate}) =>
  <RelativePositionSelect
    name="relativePosition"
    value={value}
    onChange={onChange}
    onBlur={onUpdate}
    data-test="relative-position-select"
  >
    {RELATIVE_POSITIONS.map(position => (
      <option key={position} value={position}>
        {position}
      </option>
    ))}
  </RelativePositionSelect>;

PositionPicker.propTypes = {
    value: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    onUpdate: PropTypes.func.isRequired
};

export default PositionPicker;
