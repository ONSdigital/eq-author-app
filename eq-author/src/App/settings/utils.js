export const getThemeSettingsErrorCount = (questionnaire) =>
  questionnaire?.themeSettings?.validationErrorInfo?.totalCount +
    questionnaire?.validationErrorInfo?.totalCount || 0;
