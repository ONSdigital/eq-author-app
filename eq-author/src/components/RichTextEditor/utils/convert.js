import { convertToHTML, convertFromHTML } from "draft-convert";

export const toHTML = (entityMap) => {
  const entityToHTML = (entity, originalText) => {
    const mapper = entityMap[entity.type];
    return mapper ? mapper(entity) : originalText;
  };

  const convert = convertToHTML({ entityToHTML });

  return (editorState) => convert(editorState.getCurrentContent());
};

export const fromHTML = (nodeToFn) => {
  const htmlToEntity = (nodeName, ...otherArgs) => {
    const entity = Object.entries(nodeToFn)
      .filter(([name]) => name === nodeName)
      .map(([, fn]) => fn(nodeName, ...otherArgs))
      .find((result) => result);

    return entity || null;
  };

  return convertFromHTML({ htmlToEntity });
};
