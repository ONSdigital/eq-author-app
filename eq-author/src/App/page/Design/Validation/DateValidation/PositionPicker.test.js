import React from "react";
import { shallow } from "enzyme";
import { byTestAttr } from "tests/utils/selectors";

import PositionPicker from "./PositionPicker";

describe("Date Position Picker", () => {
    let props, wrapper;
    beforeEach(() => {
        props = {
            value: "before",
            onChange: jest.fn(),
            onUpdate: jest.fn()
        };

        wrapper = shallow(<PositionPicker {...props} />);
    });

    it("Should render", () => {
        expect(wrapper).toMatchSnapshot();
    });

    it("should trigger update answer validation when the relative position changes", () => {
        const relativePositionField = wrapper.find(
            byTestAttr("relative-position-select")
        );
        relativePositionField.simulate("change", "event");
        expect(props.onChange).toHaveBeenCalledWith("event");
        relativePositionField.simulate("blur", "event");
        expect(props.onUpdate).toHaveBeenCalledWith("event");
    });
});
