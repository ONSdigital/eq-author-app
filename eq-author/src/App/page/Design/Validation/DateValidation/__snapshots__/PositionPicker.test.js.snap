// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Date Position Picker Should render 1`] = `
<components__RelativePositionSelect
  data-test="relative-position-select"
  name="relativePosition"
  onBlur={[MockFunction]}
  onChange={[MockFunction]}
  value="before"
>
  <option
    key="before"
    value="before"
  >
    before
  </option>
  <option
    key="after"
    value="after"
  >
    after
  </option>
</components__RelativePositionSelect>
`;
